File: packages/swift-client/Sources/PyricFirestore/Public/FieldValue.swift
import Foundation

/// Sentinel values that can be used when writing document fields with setData() or updateData().
public class FieldValue: Equatable, @unchecked Sendable {
    internal let sentinel: any PyricSentinel

    internal init(sentinel: any PyricSentinel) {
        self.sentinel = sentinel
    }

    public static func serverTimestamp() -> FieldValue {
        FieldValue(sentinel: PyricServerTimestampSentinel())
    }

    public static func delete() -> FieldValue {
        FieldValue(sentinel: PyricDeleteFieldSentinel())
    }

    public static func increment(_ n: Double) -> FieldValue {
        FieldValue(sentinel: PyricIncrementSentinel(n))
    }

    public static func increment(_ n: Int64) -> FieldValue {
        FieldValue(sentinel: PyricIncrementSentinel(n))
    }

    public static func arrayUnion(_ elements: [Any]) -> FieldValue {
        FieldValue(sentinel: PyricArrayUnionSentinel(elements))
    }

    public static func arrayRemove(_ elements: [Any]) -> FieldValue {
        FieldValue(sentinel: PyricArrayRemoveSentinel(elements))
    }

    public static func vector(_ values: [Double]) -> VectorValue {
        VectorValue(values)
    }

    public static func == (lhs: FieldValue, rhs: FieldValue) -> Bool {
        lhs.sentinel.isEqual(to: rhs.sentinel)
    }
}

/// A dense vector embedding representation for vector search and similarity queries.
public struct VectorValue: Equatable, Hashable, Sendable {
    public let values: [Double]

    public init(_ values: [Double]) {
        self.values = values
    }

    public var arrayValue: [Double] {
        values
    }
}

File: packages/kt-client/src/main/kotlin/com/google/firebase/firestore/FieldValue.kt
package com.google.firebase.firestore

import dev.pyric.codecs.SentinelValidator

/**
 * Sentinel values that can be used when writing document fields with set() or update().
 */
abstract class FieldValue internal constructor() {

    internal class ServerTimestampSentinel : FieldValue() {
        override fun toString(): String = "FieldValue.serverTimestamp()"
    }

    internal class DeleteSentinel : FieldValue() {
        override fun toString(): String = "FieldValue.delete()"
    }

    internal class IncrementSentinel(val operand: Number) : FieldValue() {
        override fun toString(): String = "FieldValue.increment($operand)"
    }

    internal class ArrayUnionSentinel(val elements: List<Any?>) : FieldValue() {
        override fun toString(): String = "FieldValue.arrayUnion($elements)"
    }

    internal class ArrayRemoveSentinel(val elements: List<Any?>) : FieldValue() {
        override fun toString(): String = "FieldValue.arrayRemove($elements)"
    }

    companion object {
        private val SERVER_TIMESTAMP_INSTANCE = ServerTimestampSentinel()
        private val DELETE_INSTANCE = DeleteSentinel()

        fun serverTimestamp(): FieldValue = SERVER_TIMESTAMP_INSTANCE

        fun delete(): FieldValue = DELETE_INSTANCE

        fun increment(l: Long): FieldValue = IncrementSentinel(l)

        fun increment(d: Double): FieldValue = IncrementSentinel(d)

        fun arrayUnion(vararg elements: Any?): FieldValue {
            SentinelValidator.validateNoArraySentinels(elements, "arrayUnion")
            return ArrayUnionSentinel(elements.toList())
        }

        fun arrayRemove(vararg elements: Any?): FieldValue {
            SentinelValidator.validateNoArraySentinels(elements, "arrayRemove")
            return ArrayRemoveSentinel(elements.toList())
        }
    }
}
